Add destroyMailboxes method to support destroying multiple mailboxes in a single request

sngo25 opened the pull request:
- [x] Unit test

ddolcimascolo commented on CHANGELOG.md:
```diff
@@ -3,6 +3,8 @@ All notable changes to this project will be documented in this file.
 
 ## [master]
+### Added
+- Cliennt.destroyMailboxes
```
Double N :)

sngo25 replied:
Nice catch. Done.

ddolcimascolo commented on lib/client/Client.js:
```diff
@@ -356,6 +356,32 @@ export default class Client {
+
+          throw new Error('Failed to destroy ' + id + ', the reason is: ' + (response.notDestroyed[id] || 'missing'));
+        }
+      });
```
Such bulk methods do return the whole set response object. It's not up to the library to decide if the call failed when 'some' the of the deletion failed.. 
Maybe there are business rules in the caller that accept some failures as a success, maybe the caller doesn't care, etc. 

Just remove the then block and check success in OP.

sngo25 replied:
Done.

ddolcimascolo commented on lib/client/Client.js:
```diff
@@ -356,6 +356,32 @@ export default class Client {
+   *
+   * @see http://jmap.io/spec.html#destroying-mailboxes
+   */
+  destroyMailboxes(ids) {
```
You can now make destroyMailbox delegate to destroyMailboxes

sngo25 replied:
done

ddolcimascolo commented on lib/client/Client.js:
```diff
@@ -356,6 +356,32 @@ export default class Client {
+   * Destroy multiple {@link Mailbox}es specified to the given _ids_ on the server.<br />
+   * This will issue a {@link Client#setMailboxes} JMAP request under the hoods, passing the correct options.
+   *
+   * @param ids {Array} An array IDs of the mailboxes to destroy. These IDs must be in the right order: Destroy X comes before destroy Y if X is a descendent of Y.
```
You can use String[] as the type

dbenchi commented on lib/client/Client.js:
```diff
@@ -356,6 +356,32 @@ export default class Client {
   }
 
   /**
+   * Destroy multiple {@link Mailbox}es specified to the given _ids_ on the server.<br />
```
IMO, I think that we should mention the fact that ids' order is important where we CANNOT delete a parent before deleting its parent.

sngo25 replied:
There is `These IDs must be in the right order...` and the API consumer can check http://jmap.io/spec.html#destroying-mailboxes to know the reason

dbenchi commented on CHANGELOG.md:
```diff
@@ -3,6 +3,8 @@ All notable changes to this project will be documented in this file.
 This project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [master]
```
you should indicate the bump version number

rpignolet replied:
Why ? It will be done during the release no (or just before the release) ?

ddolcimascolo replied:
No we do this after the release

dbenchi commented:
LGTM

ddolcimascolo commented on lib/client/Client.js:
```diff
@@ -336,23 +336,33 @@ export default class Client {
-        if (response.destroyed.indexOf(id) < 0) {
-          throw new Error('Failed to destroy ' + id + ', the reason is: ' + (response.notDestroyed[id] || 'missing'));
-        }
-      });
```
Please don't !
We do want to keep this logic here. Also you did factor exactly the part you should have not :)

Have a look at how destroyMessage delegates to destroyMessages please

ddolcimascolo commented on lib/client/Client.js:
```diff
@@ -336,23 +336,33 @@ export default class Client {
-   * @see http://jmap.io/spec.html#destroying-mailboxes
    */
   destroyMailbox(id) {
     Utils.assertRequiredParameterIsPresent(id, 'id');
```
This should be removed as it will be handled by destroyMessages

ddolcimascolo commented on lib/client/Client.js:
```diff
@@ -336,23 +336,33 @@ export default class Client {
-          throw new Error('Failed to destroy ' + id + ', the reason is: ' + (response.notDestroyed[id] || 'missing'));
-        }
-      });
+    return this.destroyMailboxes([id]);
```
You must do the `id && [id]` trick, or you might end up with [null] in destroyMailboxes which will pass your validation.
Also you probably lack unit tests on that. Again look at how destroyMessage delegates to destroyMessages

sngo25 replied:
Your trick will lead to wrong error if the `id` is undefined or null.

ddolcimascolo replied:
I'm ok to keep a validation in this method too, and don't use the trick. No problem
It's just that i like factoring you know :)

sngo25 commented on lib/client/Client.js:
```diff
@@ -336,23 +336,47 @@ export default class Client {
+    }).then(response => {
+      let notDestroyedIds = Object.keys(response.notDestroyed);
+
+      if (notDestroyedIds.length > 0) {
```
The promise is only resolved if all mailboxes were destroyed, if the caller wants more control, he should issue a setMailboxes request directly.

ddolcimascolo replied:
You finally did as you wished... I don't want to block this forever so I'll approve, but you're aware that as soon as we want to update the UI in OpenPaas we'll need to use setMailboxes and this helper method will become useless...

I created https://ci.open-paas.org/jira/browse/INBOX-377

sngo25 replied:
> but you're aware that as soon as we want to update the UI in OpenPaas we'll need to use setMailboxes

This is normal. In fact, if the function just do `return this.setMailboxes(destroy: ...)`, it's a likely useless helper IMO.

sngo25 committed:
Add Client#destroyMailboxes method

- Support destroying multiple mailboxes in a single request
- The promise is only resolved if all mailboxes were destroyed, if the caller
wants more control, he should issue a setMailboxes request directly

ddolcimascolo commented on lib/client/Client.js:
```diff
@@ -336,23 +336,45 @@ export default class Client {
+   *
+   * @param ids {String[]} An array IDs of the mailboxes to destroy. These IDs must be in the right order: Destroy X comes before destroy Y if X is a descendent of Y.
+   *
+   * @return {Promise} A {@link Promise} that eventually resolves to nothing if all {@link Mailbox}es were destroyed successfully.
```
You should mention that it's rejected with the first notDeleted mailbox otherwise.

sngo25 replied:
Done.

dbenchi commented:
I am not completely ok with this PR cause I prefered having a list of notDeleted folder rather than rejecting but it is ok cause I am sure that this helper will not be used widely in our code
